src/System.Collections.Concurrent/tests/ProducerConsumerCollectionTests.cs

```diff
@@ -761,8 +768,11 @@ public void ManyConcurrentAddsTakes_CollectionRemainsConsistent(int operations)
                 // avoid 0 as default(T), to detect accidentally reading a default value
                 for (int i = 1; i < (1 + operations); i++)
                 {
-                    Assert.True(c.TryAdd(i));
-                    added.Add(i);
+                    if (c.Count < MaxConcurrentCollectionSize)
+                    {
+                        Assert.True(c.TryAdd(i));
+                        added.Add(i);
+                    }
```

Comment:
That's better, but it still introduces more synchronization which will lead to more serialization.

How about just changing the test to have N threads all adding and taking, e.g. every thread picks some random number from 1 to 100 and then sits in a loop like:
```C#
for (int i = 0; i < iter; i++)
{
    for (int j = 0; j < addCount; j++) TryAdd(...);
    for (int j = 0; j < addCount; j++) TryTake(...);
}
```
Not quite the same thing being tested, but probably good enough.